src/gpu/intel/jit/gemm/gen_gemm_kernel.cpp

```diff
@@ -573,31 +573,16 @@ status_t gen_gemm_nocopy_kernel_desc_t::select_kernel(compute::gpu_arch_t arch,
         if (dt.isInt4()) return "[FO]";
         return nullptr;
     });
-
-    if (fpmath_bf16
-            && (utils::one_of(Type::f32, problem_.Ta, problem_.Tb)
-                    || (problem_.Ta.isF8() || problem_.Tb.isF8()))
-            && (problem_.Ta.isInteger() || problem_.Tb.isInteger())) {
-        if (problem_.Ta.isInt8() || problem_.Ta.isInt4()) {
-            match_params.emplace_back(match_params[0]);
-            match_params.back().selector.precisions[1] = "B";
-        } else {
-            match_params.emplace_back(match_params[0]);
-            match_params.back().selector.precisions[0] = "B";
-        }
-    }
-
-    if (fpmath_f16
-            && (utils::one_of(Type::f32, problem_.Ta, problem_.Tb)
-                    || (problem_.Ta.isF8() || problem_.Tb.isF8()))
+    if (((problem_.Ta.isF8() || problem_.Tb.isF8()))
             && (problem_.Ta.isInteger() || problem_.Tb.isInteger())) {
-        if (problem_.Ta.isInt8() || problem_.Ta.isInt4()) {
-            match_params.emplace_back(match_params[0]);
-            match_params.back().selector.precisions[1] = "H";
-        } else {
-            match_params.emplace_back(match_params[0]);
-            match_params.back().selector.precisions[0] = "H";
-        }
+        add_mode_matches(fpmath_bf16, [](Type dt) -> const char * {
+            if (dt.isF8()) return "B";
+            return nullptr;
+        });
+        add_mode_matches(fpmath_f16, [](Type dt) -> const char * {
+            if (dt.isF8()) return "H";
+            return nullptr;
```

Comment:
You should be able to just add these if statements into the corresponding `add_mode_matches` at lines 560 and 566. Is there a reason that doesn't work?